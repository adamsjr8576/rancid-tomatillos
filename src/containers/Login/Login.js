import React, { Component }from 'react';
import './Login.scss';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { addUser, updateLoggedIn, updateUserRatings } from '../../actions/index';
import { fetchUser, fetchRatings } from '../../apiCalls';


export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      error: false,
      background: ''
     }
  }


  componentDidMount = () => {
    const { movies } = this.props;
    const index = Math.floor(Math.random() * Math.floor(20));
    const imageUrl = movies.movies[index].backdrop_path;
    this.setState({background: imageUrl});
  }

  handleChange = (e) => {
    this.setState({[e.target.name]: e.target.value})
  }

  handleSubmit = () => {
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(this.state)
    }

    fetchUser(options)
      .then(data => {
        this.props.addUser(data.user.id)
        this.props.updateLoggedIn(this.props.isLoggedIn)
        return data.user.id
      })
      .then(id => {
        fetchRatings(id)
          .then(data => this.props.updateUserRatings(data.ratings))
          .catch(err => console.log(err))
      })
    .catch(error => {
      this.setState({error: true})
    })
  }



  render() {
    const {movies, loggedInStatus} = this.props
    if (loggedInStatus) {
      return (
    <Redirect to='/' />
      )
    }
    return (
      <main className='main-login' style={{backgroundImage: `url(${this.state.background})` }}>
        <form className='login-form'>
          <label htmlFor='email' className="form-label">Email</label>
          <input id='email' className="form-input" type="text" name="email"
          value={this.state.email} onChange={(e) => this.handleChange(e)} placeholder="User@turing.io" />
          <label htmlFor='password' className="form-label">Password</label>
          <input id='password' className="form-input" type="password" name="password"
          value={this.state.password} onChange={(e) => this.handleChange(e)} placeholder="password123" autocomplete='off' />
          <button onClick={this.handleSubmit} type='button' className='form-btn' id='submit-btn'>Submit</button>
          {this.state.error && <p className='error-p'>You have entered an invalid username or password, please try again</p>}
        </form>
      </main>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addUser: user => dispatch( addUser(user) ),
  updateLoggedIn: isLoggedIn => dispatch( updateLoggedIn(isLoggedIn) ),
  updateUserRatings: ratings => dispatch( updateUserRatings(ratings) )
})

export const mapStateToProps = state => ({
  loggedInStatus: state.isLoggedIn,
  movies: state.movies
})

export default connect(mapStateToProps, mapDispatchToProps)(Login);
