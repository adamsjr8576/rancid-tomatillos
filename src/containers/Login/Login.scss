.main-login {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 70vh;
  background-size: cover;
  background-position: top;
}

.login-form {
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  align-items: center;
  height: 55%;
  width: 35%;
  border: 2px solid black;
  border-radius: 10px;
  background-color: rgba(0, 0, 0, .85);
  padding: 35px;

  label {
    color: white;
    cursor: pointer;
  }

  .form-input {
    width: 75%;
    font-size: 1em;
    height: 25px;
    padding: 7px;
    margin-bottom: 10px;
    border-radius: 5px;
    border: none;
  }

  .form-btn {
  width: 50%;
  font-size: 1.5em;
  padding: 5px;
  border-radius: 10px;
  border: 0;
  background-color: white;
  color: black;
  cursor: pointer;
  outline: 0;
  margin: 20px auto;
  }

  .form-btn:hover {
    transform: scale(1.05);
  }

  .error-p {
    color: firebrick;
  }
}
