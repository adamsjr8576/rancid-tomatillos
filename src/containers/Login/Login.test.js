import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapDispatchToProps, mapStateToProps } from '../Login/Login';
import { addUser, updateLoggedIn, updateUserRatings } from '../../actions';


describe('Login Component', () => {
  let wrapper, mockAddUser, mockUpdateLoggedIn, mockUpdateUserRatings, mockDispatch, id, mockMovies;
  beforeEach(() => {
    const mockMath = Object.create(global.Math);
    mockMath.random = () => 0;
    global.Math = mockMath;
    mockMovies = { movies: [{
      id:1,
      title:"Jumanji: The Next Level",
      poster_path:"https://image.tmdb.org/t/p/original//l4iknLOenijaB85Zyb5SxH1gGz8.jpg",
      backdrop_path:"https://image.tmdb.org/t/p/original//zTxHf9iIOCqRbxvl8W5QYKrsMLq.jpg",
      release_date:"2019-12-04",
      overview:"In Jumanji: The Next Level",
      average_rating:6
    }]};
    mockAddUser = jest.fn();
    mockUpdateLoggedIn = jest.fn(() => 6);
    mockUpdateUserRatings = jest.fn();
    mockDispatch = jest.fn();
    id = 2;
    wrapper = shallow(<Login
      id={id}
      addUser = {mockAddUser}
      updateLoggedIn = {mockUpdateLoggedIn}
      updateUserRatings = {mockUpdateUserRatings}
      movies = {mockMovies}
    />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  it('should have a default state', () => {
    let mockDefaultState = {
      email: '',
      password: '',
      error: false,
      background: 'https://image.tmdb.org/t/p/original//zTxHf9iIOCqRbxvl8W5QYKrsMLq.jpg'
    }

    expect(wrapper.state()).toEqual(mockDefaultState);
  });

  it('when handleChange is invoked it should setState', () => {
    let mockEvent = {
      target: {
      name: 'email',
      value: 'Charlie@turing.io'
      }
    };
    let mockState = {
      email: 'Charlie@turing.io',
      password: '',
      error: false,
      background: 'https://image.tmdb.org/t/p/original//zTxHf9iIOCqRbxvl8W5QYKrsMLq.jpg'
    }

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state()).toEqual(mockState)
  });

  it('should invoke handleSubmit when button is clicked', async () => {
    const apiCalls = require('../../apiCalls');
    const mockRatings = {ratings: [
      {
        id: 22,
        user_id: 6,
        movie_id: 4,
        rating: 10,
        created_at: "2019-12",
        updated_at: "2019-12"
      }]
    };

    const spyUser = jest.spyOn(apiCalls, 'fetchUser').mockImplementation(() => 6);
    const spyRatings = jest.spyOn(apiCalls, 'fetchRatings').mockImplementation(() => mockRatings);
    wrapper.instance().addUser = jest.fn();
    wrapper.instance().updateLoggedIn = jest.fn();
    wrapper.instance().updateUserRatings = jest.fn();


    spyUser.mockResolvedValue();
    spyRatings.mockResolvedValue();

    wrapper.find('#submit-btn').simulate('click');

    await Promise.resolve();

    expect(spyUser).toHaveBeenCalledTimes(1);
    // expect(spyRatings).toHaveBeenCalledTimes(1);
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with addUser', () => {
      let mappedProps = mapDispatchToProps(mockDispatch);
      let actionToDispatch = addUser({ id: 6 })
      mappedProps.addUser({ id: 6} )
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with updateLoggedIn', () => {
      let isLoggedIn = true;

      let mappedProps = mapDispatchToProps(mockDispatch);
      let actionToDispatch = updateLoggedIn( isLoggedIn )
      mappedProps.updateLoggedIn(isLoggedIn)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with updateUserRatings', () => {
      let mockRatings = [
        {
          created_at: "2019-12-19T21:47:31.647Z",
          id: 22,
          movie_id: 4,
          rating: 10,
          updated_at: "2019-12-19T21:47:31.647Z",
          user_id: 6
        },
        {
          created_at: "2019-12-19T22:05:16.098Z",
          id: 23,
          movie_id: 15,
          rating: 8,
          updated_at: "2019-12-19T22:05:16.098Z",
          user_id: 6
        }
      ];
      let mappedProps = mapDispatchToProps(mockDispatch);
      let actionToDispatch = updateUserRatings(mockRatings);
      mappedProps.updateUserRatings(mockRatings);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

  describe('mapStateToProps', () => {
    const mockState = {
      isLoggedIn: false,
      movies: ['Frozen 2', 'Add Astra', 'Pee Wees Big Adventure'],
      userRatings: [],
      userId: null
    };

    const expected = {
      loggedInStatus: false,
      movies: ['Frozen 2', 'Add Astra', 'Pee Wees Big Adventure']
    }

    it('should return an object from the store containing the user is logged in and get the movies', () => {
      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected);
    });

  });
});
