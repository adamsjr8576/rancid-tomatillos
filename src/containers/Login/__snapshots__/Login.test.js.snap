// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should match the snapshot 1`] = `
<main
  className="main-login"
  style={
    Object {
      "backgroundImage": "url(https://image.tmdb.org/t/p/original//zTxHf9iIOCqRbxvl8W5QYKrsMLq.jpg)",
    }
  }
>
  <form
    className="login-form"
  >
    <label
      className="form-label"
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="form-input"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="User@turing.io"
      type="text"
      value=""
    />
    <label
      className="form-label"
      htmlFor="password"
    >
      Password
    </label>
    <input
      autocomplete="off"
      className="form-input"
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="password123"
      type="password"
      value=""
    />
    <button
      className="form-btn"
      id="submit-btn"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </form>
</main>
`;
